api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,240 @@
+			utils.Logger().Info().
+				Msgf("[EPOCHSYNC] Node is now IN SYNC! (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 60
+		}
+
+		utils.Logger().Info().
+			Msgf("[EPOCHSYNC] Node is OUT OF SYNC (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+				isBeacon, bc.ShardID(), maxHeight, height)
+
+		var heights []uint64
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
+			}
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
+		}
+
+		if len(heights) == 0 {
+			// We currently on top.
+			return 10
+		}
+
+		err := ss.ProcessStateSync(heights, bc, worker)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Msgf("[EPOCHSYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 2
+		}
+	}
+}
+
+// ProcessStateSync processes state sync from the blocks received but not yet processed so far
+func (ss *EpochSync) ProcessStateSync(heights []uint64, bc *core.BlockChain, worker *worker.Worker) error {
+
+	var payload [][]byte
+	var peerCfg *SyncPeerConfig
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		resp := peerConfig.GetClient().GetBlocksByHeights(heights)
+		if resp == nil {
+			return false
+		}
+		if len(resp.Payload) == 0 {
+			return false
+		}
+		payload = resp.Payload
+		peerCfg = peerConfig
+		return false
+	})
+
+	if len(payload) == 0 {
+		return errors.New("empty payload")
+	}
+	err := ss.processWithPayload(payload, bc)
+	if err != nil {
+		// Assume that node sent us invalid data.
+		ss.syncConfig.RemovePeer(peerCfg)
+		utils.Logger().Error().Err(err).
+			Msgf("[EPOCHSYNC] Removing peer %s for invalid data", peerCfg.String())
+		return err
+	}
+	return nil
+}
+
+func (ss *EpochSync) processWithPayload(payload [][]byte, bc *core.BlockChain) error {
+	decoded := make([]*types.Block, 0, len(payload))
+	for idx, blockBytes := range payload {
+		block, err := RlpDecodeBlockOrBlockWithSig(blockBytes)
+		if err != nil {
+			return errors.Wrap(err, "failed decode")
+		}
+
+		if !block.IsLastBlockInEpoch() {
+			return errors.Errorf("block is not last index %d hash %s", idx, block.Hash().Hex())
+		}
+
+		decoded = append(decoded, block)
+	}
+
+	for _, block := range decoded {
+		sig, bitmap, err := chain.ParseCommitSigAndBitmap(block.GetCurrentCommitSig())
+		if err != nil {
+			return errors.Wrap(err, "parse commitSigAndBitmap")
+		}
+
+		// Signature validation.
+		err = bc.Engine().VerifyHeaderSignature(bc, block.Header(), sig, bitmap)
+		if err != nil {
+			return errors.Wrap(err, "failed signature validation")
+		}
+
+		ep := block.Header().Epoch()
+		next := ep.Add(ep, common.Big1)
```

Comment:
Prefer to use the epoch within the shardState. If it doesn't exist, then use epoch + 1.